internal/jimm/role/role.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 Canonical.
+package role
+
+import (
+	"context"
+
+	"github.com/canonical/jimm/v3/internal/common/pagination"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/openfga"
+	jimmnames "github.com/canonical/jimm/v3/pkg/names"
+)
+
+// RoleStore defines a storage interface to persist roles.
+type RoleStore interface {
+	// AddRole adds a new role.
+	AddRole(ctx context.Context, name string) (re *dbmodel.RoleEntry, err error)
+
+	// GetRole populates the provided *dbmodel.RoleEntry based on name or UUID.
+	GetRole(ctx context.Context, role *dbmodel.RoleEntry) (err error)
+
+	// UpdateRoleName updates the name of a role identified by UUID.
+	UpdateRoleName(ctx context.Context, uuid, name string) (err error)
+
+	// RemoveRole removes the role identified by its ID or UUID.
+	RemoveRole(ctx context.Context, role *dbmodel.RoleEntry) (err error)
+
+	// ListRoles returns a paginated list of Roles defined by limit and offset.
+	// match is used to fuzzy find based on entries' name or uuid using the LIKE operator (ex. LIKE %<match>%).
+	ListRoles(ctx context.Context, limit, offset int, match string) (_ []dbmodel.RoleEntry, err error)
+
+	// CountRoles returns a count of the number of Roles that exist.
+	CountRoles(ctx context.Context) (count int, err error)
+}
+
+// RoleAuthorisationService holds the interface to manage roles within an authorisation service.
+type RoleAuthorisationService interface {
+	// RemoveRole removes a role.
+	RemoveRole(ctx context.Context, role jimmnames.RoleTag) error
+}
+
+// roleManager provides a means to manage roles within JIMM.
+type roleManager struct {
+	store   RoleStore
+	authSvc RoleAuthorisationService
+}
+
```

Comment:
thinking out loud:
this is good, very interfacy, i am just worried the boilerplate we are imposing upon us with this new pattern. 

I would have love to discuss this new pattern in a separate PR

Reply:
This is generally the best way to write go:
- never return interfaces
- create interfaces on consuming side to determine what methods are accessed within your service